README.md
oneko
=====

The ever popular kitty-that-follows-your-mouse-pointer toy.


Neko is Japanese for cat. According to  [Wikipedia](https://en.wikipedia.org/wiki/Neko_(software)) it was originally written for the NEC PC-9801.

There are many many versions. Generally it is considered Public Domain software, so I am placing my fork under the Unlicense.

This fork is based on [Werner Randelshofer's Java port](http://www.randelshofer.ch/blog/2010/07/screenmate-neko-in-java).

The image files are ancient public domain.
The Java code for v.1.0 is (c) 2010 Werner Randelhofer.
The Java code for v.2.0 (c) 2019 Gerald Reno

links
=====

[Download](https://glreno.github.io/oneko)
[Project wiki](https://github.com/glreno/oneko/wiki)



CHANGES.md
oneko
=====

# Change log

## V2.0.0, January 2019
* Restructured Werner Randelshofer's v.1.0
* Added support for multiple monitors.
* Added new "window mode" -- click on cat to move it into/outof a window.
* Cleaned up animation

## V2.0.1, February 2019
* Added mouse offset
* Fixed defect when un-miminizing window mode.


src/main/java/Settings.java
/*
 * @(#)Settings.java  2.0.01  2019-02-02
 *
 * Copyright (c) 2019 Jerry Reno
 * This is public domain software, under the terms of the UNLICENSE
 * http://unlicense.org 
 *
 */

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Path;
import java.util.Properties;

/** Simple file-overridable resources */
public class Settings {
	private String fn;
	private File homefile;
	private Properties builtin;
	private Properties override;

	/** Load all of the properties in src/main/resources/filename
	 * and the file ~/filename
	 */
	public Settings(String filename)
	{
		fn=filename;
		builtin=new Properties();
		override=new Properties();
		String home=System.getProperty("user.home");
		homefile=new File(home,fn);
	}

	public void load()
	{
		builtin.clear();
		InputStream in=null;
		try
		{
			in=Settings.class.getResourceAsStream(fn);
			builtin.load(in);
		}
		catch (IOException e)
		{
			e.printStackTrace();
		}
		finally
		{
			if ( in!=null ) {
				try{
				in.close();
				} catch(IOException e) {}
			}
		}

		loadOverride();
	}

	public void loadOverride()
	{
		override.clear();
		if ( homefile==null || !homefile.exists() ) return;
		InputStream in=null;
		try
		{
			in=new FileInputStream(homefile);
			override.load(in);
		}
		catch (IOException e)
		{
			e.printStackTrace();
		}
		finally
		{
			if ( in!=null ) {
				try{
				in.close();
				} catch(IOException e) {}
			}
		}
	}

	public String getString(String key)
	{
		return getString(key,null);
	}

	public String getString(String key,String def)
	{
		Object ret=null;
		ret=override.get(key);
		if ( ret!=null ) return ret.toString();
		ret=builtin.get(key);
		if ( ret!=null ) return ret.toString();
		return def;
	}

	public Integer getInt(String key)
	{
		return getInt(key,null);
	}

	public Integer getInt(String key, Integer def)
	{
		Object ret=null;
		ret=override.get(key);
		if ( ret!=null )
		{
			if ( ret instanceof Integer ) return (Integer) ret;
			try {
				Integer i = Integer.valueOf(ret.toString());
				override.put(key,i);
				return i;
			}
			catch (NumberFormatException e) { }
		}
		ret=builtin.get(key);
		if ( ret!=null )
		{
			if ( ret instanceof Integer ) return (Integer) ret;
			try {
				Integer i = Integer.valueOf(ret.toString());
				override.put(key,i);
				return i;
			}
			catch (NumberFormatException e) { }
		}
		return def;
	}

}


src/main/java/NekoSettings.java
/*
 * @(#)NekoSettings.java  2.0.1  2019-02-02
 *
 * Copyright (c) 2019 Jerry Reno
 * This is public domain software, under the terms of the UNLICENSE
 * http://unlicense.org 
 */

import java.awt.BorderLayout;
import java.awt.Color;
import java.awt.Component;
import java.awt.Container;
import java.awt.Dimension;
import java.awt.EventQueue;
import java.awt.GraphicsConfiguration;
import java.awt.GraphicsDevice;
import java.awt.Insets;
import java.awt.LayoutManager;
import java.awt.MouseInfo;
import java.awt.Point;
import java.awt.PointerInfo;
import java.awt.Rectangle;
import java.awt.Toolkit;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.event.MouseEvent;
import java.awt.event.MouseListener;
import javax.swing.ImageIcon;
import javax.swing.JFrame;
import javax.swing.JLabel;
import javax.swing.JWindow;
import javax.swing.Timer;
import javax.swing.WindowConstants;

public class NekoSettings {

	// Settings keys:
	private static final String HELLO = "hello";
	private static final String TITLE = "windowTitle";

	private static final String TRIGGER_DIST = "triggerDistance";
	private static final String CATCH_DIST = "catchDistance";
	private static final String RUN_DIST = "runDistancePerFrame";
	private static final String OFFSETX = "offsetx";
	private static final String OFFSETY = "offsety";

	private static final String MAX_FRAMERATE = "maxFramerate";
	private static final String RUN_FRAMERATE = "runFramerate";
	private static final String SIT_FRAMERATE = "sitFramerate";
	private static final String SHARPEN_FRAMERATE = "sharpenFramerate";
	private static final String SCRATCH_FRAMERATE = "scratchFramerate";
	private static final String LOAD_FRAMERATE = "loadFramerate";

	private static final String SLEEP_DELAY = "sleepDelay";
	private static final String YAWN_DELAY = "yawnDelay";
	private static final String SURPRISE_DELAY = "surpriseDelay";

	private Settings settings;
	private int triggerDist;
	private int catchDist;
	private int runDist;
	private int minDelay;
	private int runDelay;
	private int sitDelay;
	private int scratchDelay;
	private int sharpenDelay;
	private int loadDelay;
	private int sleepDelay;
	private int yawnDelay;
	private int surpriseDelay;
	private int offsetX;
	private int offsetY;

	public NekoSettings() {
		settings=new Settings("neko.properties");
		load();
	}

	public String getTitle() { return settings.getString(TITLE);}

	/** Convert frames-per-seconds to milliseconds */
	private int getDelay(String key)
	{
		// Convert the framerate settings into milliseconds
		Integer ret = settings.getInt(key,100);

		return 1000/ret;
	}

	public int getTriggerDist() { return triggerDist;}
	public int getCatchDist() { return catchDist;}
	public int getRunDist() { return runDist;}
	public int getOffsetX() { return offsetX;}
	public int getOffsetY() { return offsetY;}
	public int getMinDelay() { return minDelay;}
	public int getRunDelay() { return runDelay;}
	public int getSitDelay() { return sitDelay;}
	public int getScratchDelay() { return scratchDelay;}
	public int getSharpenDelay() { return sharpenDelay;}
	public int getLoadDelay() { return loadDelay;}
	public int getSleepDelay() { return sleepDelay;}
	public int getYawnDelay() { return yawnDelay;}
	public int getSurpriseDelay() { return surpriseDelay;}

	public void load()
	{
		settings.load();
		String hello=settings.getString(HELLO);
		if ( hello!=null ) System.out.println(hello);

		triggerDist = settings.getInt(TRIGGER_DIST,16);
		catchDist = settings.getInt(CATCH_DIST,16);
		runDist = settings.getInt(RUN_DIST,16);
		offsetX = settings.getInt(OFFSETX,0);
		offsetY = settings.getInt(OFFSETY,0);

		sleepDelay = settings.getInt(SLEEP_DELAY,1000);
		yawnDelay = settings.getInt(YAWN_DELAY,1000);
		surpriseDelay = settings.getInt(SURPRISE_DELAY,1000);

		minDelay=getDelay(MAX_FRAMERATE);
		runDelay=getDelay(RUN_FRAMERATE);
		sitDelay=getDelay(SIT_FRAMERATE);
		scratchDelay=getDelay(SCRATCH_FRAMERATE);
		sharpenDelay=getDelay(SHARPEN_FRAMERATE);
		loadDelay=getDelay(LOAD_FRAMERATE);

	}

}


src/main/java/NekoController.java
/*
 * @(#)NekoController.java  2.0.1  2019-02-02
 *
 * Copyright (c) 2019 Jerry Reno
 * This is public domain software, under the terms of the UNLICENSE
 * http://unlicense.org 
 * 
 * This is an extended version of the earlier Java Neko 1.0:
 * Copyright (c) 2010 Werner Randelshofer
 * Hausmatt 10, Immensee, CH-6405, Switzerland.
 *
 * This source code is free to everyone.
 *
 * This is a desktop adaptation of the applet
 * JAVA NEKO V1.0 by Chris Parent, 1999.
 * http://mysite.ncnetwork.net/res8t1xo/class/neko.htm
 */

import java.awt.Dimension;
import java.awt.GraphicsConfiguration;
import java.awt.GraphicsDevice;
import java.awt.Insets;
import java.awt.MouseInfo;
import java.awt.Point;
import java.awt.PointerInfo;
import java.awt.Rectangle;
import java.awt.Toolkit;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import javax.swing.ImageIcon;
import javax.swing.JFrame;
import javax.swing.JLabel;
import javax.swing.JWindow;
import javax.swing.Timer;

/**
 * Neko the cat.
 * <p>
 * This program loads in 32 images of Neko, and tests them. (to show you that
 * they've been loaded). Neko will chase you mouse cursor around the desktop.
 * Once she's over it and the mouse doesn't move she'll prepare to take a nap.
 * If the mouse go's outside the desktop she will reach the border and try
 * to dig for it. She'll eventually give up, and fall asleep.
 *
 *
 * @author Werner Randelshofer (adaption for desktop)
 *		 Chris Parent (original code)
 * @version 1.0.1 2010-07-17 Fixes timers. Sets longer sleep times when the
 * cat sleeps.
 * <br>1.0 2010-07-16 Created.
 */
public class NekoController {
	//
	//Constants

	private static final double pi = Math.PI;
	private enum Position { OVER, UNDER, LEFT, RIGHT };

	//
	//Variables
	private boolean windowMode;
	private int ox, oy;   //image pos.
	private int no;	   //image number.
	private int init;	 //for image loading initialize counter
	private int state;
	private int slp;	  //sleep time
	private int ilc1;	 //image loop counter
	private int ilc2;	 //second loop counter
	private boolean mouseMoved; //mouse move, flag
	private ImageIcon image[];  //images
	private Point windowOffset = new Point(-16, -30);
	private Rectangle nekoBounds = new Rectangle();
	private Timer timer;
	private int w,h; // size of icons

	//
	// UI Components
	private JWindow invisibleWindow;
	private JFrame catbox;
	private JLabel freeLabel,boxLabel;
	private NekoSettings settings;

	/** Creates new form Neko */
	public NekoController(NekoSettings settings,JWindow invisibleWindow,JFrame visibleWindow,JLabel free,JLabel boxed) {
		this.settings=settings;
		this.invisibleWindow=invisibleWindow;
		this.catbox=visibleWindow;
		this.freeLabel=free;
		this.boxLabel=boxed;
		this.windowMode=false;

		this.init=0;
		this.state=0;
		this.slp=0;

		loadKitten();
		w=image[1].getIconWidth();
		h=image[1].getIconHeight();
		invisibleWindow.setSize(w,h);
		//invisibleWindow.setLocation(ox + windowOffset.x, oy + windowOffset.y);
		boxLabel.setSize(w,h);
		catbox.setSize(16*w, 9*h);

		timer = new Timer(50, new ActionListener() {

			public void actionPerformed(ActionEvent e) {
				try {
					locateMouseAndAnimateCat();
				}
				catch (Exception ex) {}
			}
		});
		timer.setRepeats(true);
		timer.start();

	}

	private void loadKitten() {
		// Note: The files start with 1. image[0] is a copy of 25.
		// image[0] is a default, image[25] is part of the wash animation with image[31]
		image = new ImageIcon[33];
		for (int i = 1; i <= 32; i++) {
			image[i] = new ImageIcon(Neko.class.getResource("images/" + i + ".GIF"));
		}
		image[0]=image[25];
	}

	public int getWidth() { return w;}
	public int getHeight() { return h;}
	public boolean getWindowMode() { return this.windowMode;}

	private void calculateBounds(PointerInfo pointerInfo)
	{
		if(windowMode)
		{
			// nekoBounds is the area of the window that the Neko can be in
			Point panePoint = catbox.getContentPane().getLocationOnScreen();
			Insets paneInsets=catbox.getContentPane().getInsets();
			Dimension sz=catbox.getContentPane().getSize();

			nekoBounds.x = panePoint.x + paneInsets.left + w/2;
			nekoBounds.y = panePoint.y + paneInsets.top + h;
			nekoBounds.width = sz.width - paneInsets.left - paneInsets.right - w;
			nekoBounds.height = sz.height - paneInsets.left - paneInsets.top - h;
		}
		else
	{
			GraphicsDevice mouseMonitor=pointerInfo.getDevice();
			if (mouseMonitor==null) return;
			GraphicsConfiguration gc = mouseMonitor.getDefaultConfiguration();
			if ( gc==null) return;
			Rectangle screenBounds = gc.getBounds();
			if ( screenBounds==null) return;
			Toolkit tk = invisibleWindow.getToolkit();
			Insets screenInsets = tk.getScreenInsets(gc);
			if ( screenInsets==null) return;
			// nekoBounds is the area of the screen that the Neko can be in
			nekoBounds.x = screenBounds.x + screenInsets.left + w/2;
			nekoBounds.y = screenBounds.y + screenInsets.top + h;
			nekoBounds.width = screenBounds.width - screenInsets.left - screenInsets.right - w;
			nekoBounds.height = screenBounds.height - screenInsets.left - screenInsets.top - h;
		}
	}

	/** Locates the mouse on the screen and determines what the cat shall do. */
	public void locateMouseAndAnimateCat() {
		PointerInfo pointerInfo = MouseInfo.getPointerInfo();
		if (pointerInfo==null) return;
		Point mouseLocation = pointerInfo.getLocation();
		if (mouseLocation==null) return;

		int mx = mouseLocation.x + settings.getOffsetX();
		int my = mouseLocation.y + settings.getOffsetY();

		calculateBounds(pointerInfo);

		//Determines what the cat should do, if the mouse moves
		Position pos = null;
		boolean out = !nekoBounds.contains(mx, my);
		int x = mx;
		int y = my;
		if (out) {
			if (y < nekoBounds.y) {
				y = nekoBounds.y;
				pos = Position.OVER;
			}
			if (y > nekoBounds.y + nekoBounds.height) {
				y = nekoBounds.y + nekoBounds.height;
				pos = Position.UNDER;
			}
			if (x < nekoBounds.x) {
				x = nekoBounds.x;
				pos = Position.LEFT;
			}
			if (x > nekoBounds.x + nekoBounds.width) {
				x = nekoBounds.x + nekoBounds.width;
				pos = Position.RIGHT;
			}
		}

		//image-mouse distance
		// x,y are the mouse location on screen
		// ox,oy are the old neko location on screen
		int dx = x - ox;
		int dy = oy - y;
		double dist = Math.sqrt(dx * dx + dy * dy); //distance formula (from mouse to cat)
		double theta = Math.atan2(dy, dx);	 //angle from mouse to cat

		// If the kitty is already moving, keep moving;
		// If not moving, but the mouse is over 64px away, wake up!
		mouseMoved = mouseMoved || dist > settings.getTriggerDist();
		//
		slp = Math.max(0, slp - timer.getDelay());
		if (slp == 0)
		{
			animateCat(pos,theta,dist);
		}
	}

	private void animateCat(Position pos,double theta,double dist)
	{
		// There are four states.
		// 0. Initialization (init<33)
		// 1. Chasing the mouse (doMove=true)
		// 2. sleeping or preparing to sleep
		// 3. Surprised (was asleep, but the mouse moved) (no=32)
		boolean doMove=false;
		if ( state==0 ) {
			if (init < 33) {
				doMove=true;
				//tells the user the program is testing the images, and tells them
				//when the test is done.
				slp = settings.getLoadDelay();
				ox = nekoBounds.x + nekoBounds.width / 2;
				oy = nekoBounds.y + nekoBounds.height / 2;
				no = init;
				init++;
			}
			else {
				state=1;
			}
		}
		else if ( state==1 ) {
			doMove=true;
			slp = settings.getRunDelay(); // 5fps, 200ms
			// note that ox,oy are screen-relative
			double run=settings.getRunDist(); // 16
			if ( run>dist ) {
				run=dist;
			}
			ox = (int) (ox + Math.cos(theta) * run);
			oy = (int) (oy - Math.sin(theta) * run);
			dist = dist - run;
			if ( dist < settings.getCatchDist() ) {
				// mouse has been caught!
				state=2;
			}

			/*
			The following conditions determine what image should be shown.
			Remember there are two images for each action. For example if the cat's
			going right, display the cat with open legs and then with close legs,
			open, and so on.
			 */
			if (theta >= -pi / 8 && theta <= pi / 8) //right
			{
				no = (no == 5) ? 6 : 5;
			}
			if (theta > pi / 8 && theta < 3 * pi / 8) //upper-right
			{
				no = (no == 3) ? 4 : 3;
			}
			if (theta >= 3 * pi / 8 && theta <= 5 * pi / 8) //up
			{
				no = (no == 1) ? 2 : 1;
			}
			if (theta > 5 * pi / 8 && theta < 7 * pi / 8) //upper-left
			{
				no = (no == 15) ? 16 : 15;
			}
			if (theta >= 7 * pi / 8 || theta <= -7 * pi / 8) //left
			{
				no = (no == 13) ? 14 : 13;
			}
			if (theta > -7 * pi / 8 && theta < -5 * pi / 8) //bottom-left
			{
				no = (no == 11) ? 12 : 11;
			}
			if (theta >= -5 * pi / 8 && theta <= -3 * pi / 8) //down
			{
				no = (no == 9) ? 10 : 9;
			}
			if (theta > -3 * pi / 8 && theta < -pi / 8) //bottom-right
			{
				no = (no == 7) ? 8 : 7;
			}
			//We have moved toward the mouse, so set mouseMoved back to false
			mouseMoved = false;
		}
		else {   //-if the mouse hasn't moved or the cat's over the mouse-
			// state is 2 ; it stays at 2 until the mouse moves
			// When it moves, go to state 3, show surprise, and then to state 1
//			ox = x;
//			oy = y;
			switch (no) {
				case 0: //<cat sit>
					//If the mouse is outside the applet
					if (pos != null) {
						slp = settings.getSharpenDelay();
						switch (pos) {
							case OVER:
								no = 17;
								break;
							case UNDER:
								no = 21;
								break;
							case LEFT:
								no = 23;
								break;
							case RIGHT:
								no = 19;
								break;
							default:
								slp = settings.getSitDelay();
								no = 31;
								break;
						}
						break;
					}
					slp = settings.getSitDelay();
					no = 31;
					break; //<31: cat lick>
				//
				case 17: //The mouse is outside, above applet
					slp = settings.getSharpenDelay();
					no = 18;	//show images 17 & 18, 6 times
					ilc1++;
					if (ilc1 == 6) {
						no = 27;
						ilc1 = 0;
					}
					break;
				//
				case 18:
					slp = settings.getSharpenDelay();
					no = 17;
					break;
				//
				case 21: //The mouse is outside, under applet
					slp = settings.getSharpenDelay();
					no = 22;	//show images 21 & 22, 6 times
					ilc1++;
					if (ilc1 == 6) {
						no = 27;
						ilc1 = 0;
					}
					break;
				//
				case 22:
					slp = settings.getSharpenDelay();
					no = 21;
					break;
				//
				case 23: //the mouse is outside, left
					slp = settings.getSharpenDelay();
					no = 24;	//show images 23 & 24, 6 times
					ilc1++;
					if (ilc1 == 6) {
						no = 27;
						ilc1 = 0;
					}
					break;
				//
				case 24:
					slp = settings.getSharpenDelay();
					no = 23;
					break;
				//
				case 19: //The mouse is outside, right
					slp = settings.getSharpenDelay();
					no = 20;	//show images 19 & 20, 6 times
					ilc1++;
					if (ilc1 == 6) {
						no = 27;
						ilc1 = 0;
					}
					break;
				//
				case 20:
					slp = settings.getSharpenDelay();
					no = 19;
					break;
				//
				case 31: //cat lick (6  times)
					slp = settings.getSitDelay();
					no = 25;
					ilc1++;
					if (ilc1 == 6) {
						slp = settings.getScratchDelay();
						no = 27;
						ilc1 = 0;
					}
					break;
				//
				case 25:
					slp = settings.getSitDelay();
					no = 31;
					break;
				//
				case 27:
					slp = settings.getScratchDelay();
					no = 28;
					break; //cat scratch (27 & 28, 4 times)
				case 28:
					no = 27;
					ilc2++;
					if (ilc2 == 4) {
						no = 26;
						slp = settings.getYawnDelay();
						ilc2 = 0;
					}
					break;
				case 26:
					no = 29;
					slp = settings.getSleepDelay();
					break; //cat yawn (26)
				case 29:
					no = 30;
					slp = settings.getSleepDelay();
					break; //cat sleep (29 & 30, forever)
				case 30:
					no = 29;
					slp = settings.getSleepDelay();
					break;
				default:
					no = 0;
					break;
			}
			if (mouseMoved ) {
				// State 3. Go to state 1 and chase the mouse,
				// after a suitable delay.
				//re-initialize some variables
				slp = settings.getSurpriseDelay();
				no = 32;
				ilc1 = 0;
				ilc2 = 0;
				//mouseMoved = false;
				state=1;
			}
		}
		//draw the new image
		if ( doMove )
		{
			moveCat();
		}
		freeLabel.setIcon(image[no]);
		boxLabel.setIcon(image[no]);
	}

	private void moveCat() {
		if(windowMode) {
			// note that ox,oy are screen-relative
			boxLabel.setLocation(ox-nekoBounds.x, oy-nekoBounds.y);
		}
		else {
			invisibleWindow.setLocation(ox + windowOffset.x, oy + windowOffset.y);
		}
	}

	public void setWindowMode(boolean windowed) {
		this.windowMode=windowed;
	}

	public void moveCatInBox() {
		if ( windowMode ) {
			Dimension d=catbox.getSize();
			catbox.setLocation(ox-d.width/2,Math.max(0,oy-d.height/2));
		}
		PointerInfo pointerInfo = MouseInfo.getPointerInfo();
		if (pointerInfo==null) return;
		calculateBounds(pointerInfo);
		moveCat();
	}

	public void catboxDeiconified() {
		// Calculate the new nekoBounds
		PointerInfo pointerInfo = MouseInfo.getPointerInfo();
		if (pointerInfo==null) return;
		calculateBounds(pointerInfo);
		// Move the cat to the center of the window
		ox=nekoBounds.x+nekoBounds.width/2;
		oy=nekoBounds.y+nekoBounds.height/2;
	}

	public void catboxMoved() {
		if ( windowMode ) {
			// The window moved. We have the old location in nekoBounds.
			int oldx=nekoBounds.x;
			int oldy=nekoBounds.y;
			// Calculate the new nekoBounds
			PointerInfo pointerInfo = MouseInfo.getPointerInfo();
			if (pointerInfo==null) return;
			calculateBounds(pointerInfo);
			// How much did it move?
			int dx=nekoBounds.x-oldx;
			int dy=nekoBounds.y-oldy;
			ox+=dx;
			oy+=dy;
			// don't need to move the jlabel, it will have moved with the window
		}
	}

}



src/main/java/Neko.java
/*
 * @(#)Neko.java  2.0.1  2019-02-29
 *
 * Copyright (c) 2019 Jerry Reno
 * This is public domain software, under the terms of the UNLICENSE
 * http://unlicense.org 
 * 
 * This is an extended version of the earlier Java Neko 1.0:
 * Copyright (c) 2010 Werner Randelshofer
 * Hausmatt 10, Immensee, CH-6405, Switzerland.
 *
 * This source code is free to everyone.
 *
 * This is a desktop adaptation of the applet
 * JAVA NEKO V1.0 by Chris Parent, 1999.
 * http://mysite.ncnetwork.net/res8t1xo/class/neko.htm
 */

import java.awt.BorderLayout;
import java.awt.Color;
import java.awt.Component;
import java.awt.Container;
import java.awt.Dimension;
import java.awt.EventQueue;
import java.awt.LayoutManager;
import java.awt.Point;
import java.awt.event.ComponentAdapter;
import java.awt.event.ComponentEvent;
import java.awt.event.ComponentListener;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;
import java.awt.event.MouseListener;
import java.awt.event.WindowAdapter;
import java.awt.event.WindowEvent;
import java.awt.event.WindowListener;
import javax.swing.JFrame;
import javax.swing.JLabel;
import javax.swing.JWindow;
import javax.swing.WindowConstants;

/**
 * Neko the cat.
 * <p>
 * This program loads in 32 images of Neko, and tests them. (to show you that
 * they've been loaded). Neko will chase you mouse cursor around the desktop.
 * Once she's over it and the mouse doesn't move she'll prepare to take a nap.
 * If the mouse go's outside the desktop she will reach the border and try
 * to dig for it. She'll eventually give up, and fall asleep.
 *
 *
 * @author Werner Randelshofer (adaption for desktop)
 *		 Chris Parent (original code)
 * @version 1.0.1 2010-07-17 Fixes timers. Sets longer sleep times when the
 * cat sleeps.
 * <br>1.0 2010-07-16 Created.
 */
public class Neko {

	//
	//Constants
	private static Dimension MINSIZE=new Dimension(64,64);
	private static Dimension PRFSIZE=new Dimension(64*16,64*9);

	//
	// UI Components
	private JFrame catbox;
	private JWindow invisibleWindow;
	private JLabel freeLabel,boxLabel;
	private NekoSettings settings;
	private NekoController controller;

	/** Creates new form Neko */
	public Neko() {
		settings=new NekoSettings();
		initComponents();
		invisibleWindow.setLocation(100,100);
		controller=new NekoController(settings,invisibleWindow,catbox,freeLabel,boxLabel);
		setWindowMode(false);
	}

	/** This method is called from within the constructor to
	 * initialize the form.
	 */
	private void initComponents() {
		String title=settings.getTitle();
		if ( title==null ) title="Neko";
		catbox=new JFrame(title);
		catbox.setBackground(new Color(200,200,200,255));
		catbox.setDefaultCloseOperation(WindowConstants.EXIT_ON_CLOSE);
		invisibleWindow=new JWindow();
		invisibleWindow.getRootPane().putClientProperty("Window.shadow", false);
		invisibleWindow.setBackground(new Color(200,200,200,0)); // transparent, light grey of not supported
		invisibleWindow.setAlwaysOnTop(true);

		freeLabel = new JLabel();
		boxLabel = new JLabel();

		MouseListener mouseMoveListener = new MouseAdapter() {
			public void mouseClicked(MouseEvent evt) {
				if ((evt.getSource() == freeLabel)||(evt.getSource() == boxLabel)) {
					imageClicked(evt);
				}
			}
		};

		freeLabel.addMouseListener(mouseMoveListener);
		boxLabel.addMouseListener(mouseMoveListener);
		invisibleWindow.getContentPane().add(freeLabel, BorderLayout.CENTER);
		catbox.getContentPane().add(boxLabel);
		catbox.pack();

		// We really don't want a layout manager messing with us.
		// Maybe this class should BE a layout manager.
		catbox.getContentPane().setLayout(new LayoutManager() {
			public void addLayoutComponent(String n,Component c) {}
			public void layoutContainer(Container p){}
			public Dimension minimumLayoutSize(Container p) { return MINSIZE;}
			public Dimension preferredLayoutSize(Container p) { return PRFSIZE;}
			public void removeLayoutComponent(Component c) {}
		});

		catbox.addComponentListener(new ComponentAdapter() {
			public void componentMoved(ComponentEvent e) {
				controller.catboxMoved();
			}
		});
		catbox.addWindowListener(new WindowAdapter() {
			public void windowDeiconified(WindowEvent e) {
				controller.catboxDeiconified();
			}
		});

		invisibleWindow.pack();
	}

	private void imageClicked(MouseEvent evt) {
		setWindowMode(!controller.getWindowMode());
	}

	/**
	 * @param args the command line arguments
	 */
	public static void main(String args[]) {
		EventQueue.invokeLater(new Runnable() {

			public void run() {
				new Neko();
			}
		});
	}

	public void setWindowMode(boolean windowed)
	{
		controller.setWindowMode(windowed);
		settings.load();
		if (windowed) {
			String title=settings.getTitle();
			if ( title==null ) title="Neko";
			catbox.setTitle(title);
			invisibleWindow.setVisible(false);

			catbox.setVisible(true);
		}
		else {
			catbox.setVisible(false);
			invisibleWindow.setVisible(true);
		}
		controller.moveCatInBox();
	}

}

